extras/indexing/src/main/java/org/apache/rya/indexing/entity/EntityIndexOptimizer.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.log4j.Logger;
+import org.apache.rya.indexing.entity.model.Entity;
+import org.apache.rya.indexing.entity.model.Type;
+import org.apache.rya.indexing.entity.query.EntityQueryNode;
+import org.apache.rya.indexing.entity.storage.EntityStorage;
+import org.apache.rya.indexing.entity.storage.TypeStorage;
+import org.apache.rya.indexing.entity.update.mongo.MongoEntityIndexer;
+import org.apache.rya.indexing.external.matching.AbstractExternalSetOptimizer;
+import org.apache.rya.indexing.external.matching.ExternalSetMatcher;
+import org.apache.rya.indexing.external.matching.ExternalSetProvider;
+import org.apache.rya.indexing.external.matching.QueryNodeListRater;
+import org.apache.rya.indexing.external.matching.QuerySegment;
+import org.apache.rya.indexing.external.matching.TopOfQueryFilterRelocator;
+import org.openrdf.query.BindingSet;
+import org.openrdf.query.Dataset;
+import org.openrdf.query.algebra.TupleExpr;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Optional;
+
+/**
+ * Optimizes a query over {@link Entity}s.
+ */
+public class EntityIndexOptimizer extends AbstractExternalSetOptimizer<EntityQueryNode> implements Configurable {
+    private static final Logger log = Logger.getLogger(EntityIndexOptimizer.class);
+    private static final EntityExternalSetMatcherFactory MATCHER_FACTORY = new EntityExternalSetMatcherFactory();
+
+    private final MongoEntityIndexer indexer;
+    private EntityIndexSetProvider provider;
+    private Configuration conf;
+
+    private EntityStorage entityStorage;
+    private TypeStorage typeStorage;
+
+    /**
+     * Creates a new {@link EntityIndexOptimizer}.
+     * @param typeStorage - The mechanism for access the various {@link Type}s. (not null)
+     * @param entityStorage - The mechanism for access to the various {@link Entity}s. (not null)
+     */
+    public EntityIndexOptimizer() {
+        indexer = new MongoEntityIndexer();
+
+        //There is no rater for entity query nodes.
+        useOptimal = false;
+    }
+
+    @Override
+    public void setConf(final Configuration conf) {
+        this.conf = conf;
+        indexer.setConf(conf);
+
+        typeStorage = indexer.getTypeStorage(conf);
+        entityStorage = indexer.getEntityStorage(conf);
+
+        provider = new EntityIndexSetProvider(typeStorage, entityStorage);
+    }
+
+    @Override
+    public Configuration getConf() {
+        return conf;
+    }
+
+    @Override
+    public void optimize(TupleExpr tupleExpr, final Dataset dataset, final BindingSet bindings) {
+        checkNotNull(tupleExpr);
+        checkNotNull(indexer);
+
+        // first standardize query by pulling all filters to top of query if
+        // they exist using TopOfQueryFilterRelocator
+        tupleExpr = TopOfQueryFilterRelocator.moveFiltersToTop(tupleExpr);
+        super.optimize(tupleExpr, null, null);
+    }
+
+    @VisibleForTesting
+    public EntityStorage getEntityStorage() {
```

Comment:
This methods that only exist for testing are a smell. You should be able to push a mock or the object you need access to within your tests into this class.